### ISSUE TITLE
Public functions that are never used/tested

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1592]
This is a list of public non-deprecated functions that are never used either directly or indirectly in the main codebase or the tests/examples.

Tests should be written for some of these functions.

If you are looking to contribute to igraph, fixing some of these is an easy way to get started. If you are unsure which one to take, just ask. 

 - [ ] igraph_adjlist_fprint
 - [ ] igraph_inclist_print
 - [ ] igraph_inclist_fprint
 - [ ] igraph_arpack_unpack_complex
 - [ ] igraph_attribute_combination_remove
 - [ ] igraph_is_bipartite
 - [ ] igraph_bipartite_game
 - [ ] igraph_hub_score
 - [ ] igraph_authority_score
 - [ ] igraph_maximal_cliques_file
 - [ ] igraph_clique_size_hist
 - [ ] igraph_compare_communities
 - [ ] igraph_split_join_distance
 - [ ] igraph_le_community_to_membership
 - [ ] igraph_modularity_matrix
 - [ ] igraph_complex_csc
 - [ ] igraph_complex_cot
 - [ ] igraph_full_citation
 - [ ] igraph_extended_chordal_ring
 - [ ] igraph_linegraph
 - [ ] igraph_kautz
 - [ ] igraph_get_adjacency_sparse
 - [ ] igraph_eigen_adjacency
 - [ ] igraph_eigen_laplacian
 - [ ] igraph_laplacian_spectral_embedding
 - [ ] igraph_dim_select
 - [ ] igraph_sir
 - [ ] igraph_graphlets
 - [ ] igraph_solve_lsap
 - [ ] igraph_maximum_matching
 - [ ] igraph_motifs_randesu_estimate
 - [ ] igraph_motifs_randesu_no
 - [ ] igraph_dyad_census
 - [ ] igraph_list_triangles
 - [ ] igraph_neighborhood_size
 - [ ] igraph_neighborhood
 - [ ] igraph_neighborhood_graphs
 - [ ] igraph_local_scan_k_ecount
 - [ ] igraph_local_scan_k_ecount_them
 - [ ] igraph_local_scan_neighborhood_ecount
 - [ ] igraph_scg_norm_eps
 - [ ] igraph_stack_ptr_destroy_all
 - [ ] igraph_random_spanning_tree
 - [ ] igraph_rewire
 - [ ] igraph_unfold_tree
 - [ ] igraph_is_mutual
 - [ ] igraph_is_chordal
 - [ ] igraph_strvector_set2
 - [ ] igraph_strvector_remove
 - [ ] igraph_strvector_permdelete
 - [ ] igraph_strvector_remove_negidx
 - [ ] igraph_strvector_print
 - [ ] igraph_transitivity_avglocal_undirected
 - [ ] igraph_transitivity_barrat
 - [ ] igraph_real_snprintf
 - [ ] igraph_real_printf_precise
 - [ ] igraph_vector_floor

## Flow and connectivity

 - [ ] igraph_st_mincut
 - [ ] igraph_st_mincut_value
 - [ ] igraph_st_edge_connectivity
 - [ ] igraph_edge_disjoint_paths
 - [ ] igraph_vertex_disjoint_paths
 - [ ] igraph_adhesion
 - [ ] igraph_cohesion
 - [ ] igraph_residual_graph

## Foreign

 - [ ] igraph_write_graph_lgl
 - [ ] igraph_write_graph_dimacs

## Games

 - [ ] igraph_nonlinear_barabasi_game
 - [ ] igraph_growing_random_game
 - [ ] igraph_barabasi_aging_game
 - [ ] igraph_recent_degree_game
 - [ ] igraph_recent_degree_aging_game
 - [ ] igraph_callaway_traits_game
 - [ ] igraph_establishment_game
 - [ ] igraph_rewire_directed_edges
 - [ ] igraph_lastcit_game
 - [ ] igraph_cited_type_game
 - [ ] igraph_citing_cited_type_game
 - [ ] igraph_forest_fire_game
 - [ ] igraph_simple_interconnected_islands_game
 - [ ] igraph_static_power_law_game
 - [ ] igraph_sbm_game
 - [ ] igraph_hsbm_game
 - [ ] igraph_hsbm_list_game
 - [ ] igraph_correlated_pair_game
 - [ ] igraph_tree_game
 - [ ] igraph_dot_product_game
 - [ ] igraph_sample_sphere_volume
 - [ ] igraph_sample_dirichlet

## HRG

 - [ ] igraph_hrg_game
 - [ ] igraph_hrg_create

## Iterators

 - [ ] igraph_vss_adj
 - [ ] igraph_vs_none
 - [ ] igraph_vs_vector
 - [ ] igraph_vs_vector_copy
 - [ ] igraph_vs_type
 - [ ] igraph_es_vector
 - [ ] igraph_es_fromto
 - [ ] igraph_es_seq
 - [ ] igraph_ess_seq
 - [ ] igraph_es_path
 - [ ] igraph_es_as_vector
 - [ ] igraph_es_type

## Linalg

 - [ ] igraph_lapack_dgetrf
 - [ ] igraph_lapack_dgetrs

## Layout

 - [ ] igraph_layout_star
 - [ ] igraph_layout_fruchterman_reingold
 - [ ] igraph_layout_kamada_kawai
 - [ ] igraph_layout_springs
 - [ ] igraph_layout_reingold_tilford_circular
 - [ ] igraph_layout_random_3d
 - [ ] igraph_layout_fruchterman_reingold_3d
 - [ ] igraph_layout_graphopt
 - [ ] igraph_layout_bipartite
 - [ ] igraph_layout_drl_options_init
 - [ ] igraph_layout_drl
 - [ ] igraph_layout_drl_3d
 - [ ] igraph_layout_gem
 - [ ] igraph_layout_davidson_harel

## Non-graph

 - [ ] igraph_running_mean
 - [ ] igraph_zeroin
 - [ ] igraph_bfgs

## Paths

 - [ ] igraph_diameter_dijkstra
 - [ ] igraph_shortest_paths_johnson
 - [ ] igraph_average_path_length_dijkstra
 - [ ] igraph_path_length_hist
 - [ ] igraph_random_walk

## Error, interruption and progress handlers

- [ ] igraph_progress_handler_stderr
- [ ] igraph_set_progress_handler
- [ ] igraph_set_fatal_handler
- [ ] igraph_fatalf 
- [ ] igraph_set_interruption_handler
- [ ] igraph_status_handler_stderr
- [ ] igraph_set_status_handler

## RNG

 - [ ] igraph_rng_init
 - [ ] igraph_rng_destroy
 - [ ] igraph_rng_max
 - [ ] igraph_rng_min
 - [ ] igraph_rng_name
 - [ ] igraph_rng_set_default

## Sparsemat

 - [ ] igraph_sparsemat_fkeep
 - [ ] igraph_sparsemat_droptol
 - [ ] igraph_sparsemat_nonzero_storage
 - [ ] igraph_sparsemat_getelements_sorted
 - [ ] igraph_i_sparsemat_view (why is this public? does it need to be renamed?)
 - [ ] igraph_sparsemat_iterator_idx
 - [ ] igraph_spmatrix_add_col_values
 - [ ] igraph_spmatrix_clear_col
 - [ ] igraph_spmatrix_clear_row
 - [ ] igraph_spmatrix_max_nonzero
 - [ ] igraph_spmatrix_scale
 - [ ] igraph_spmatrix_print
 - [ ] igraph_i_spmatrix_get_col_nonzero_indices
 - [ ] igraph_i_spmatrix_clear_row_fast (why is this public?)
 - [ ] igraph_i_spmatrix_cleanup (why is this public?)

## Isomorphism

 - [ ] igraph_subisomorphic
 - [ ] igraph_get_isomorphisms_vf2
 - [ ] igraph_get_subisomorphisms_vf2

**Comment by GroteGnoom** [comment]
I would like to work on this. I looked at `igraph_st_mincut`, which only calls `igraph_maxflow`, which is already tested via `igraph_maxflow_value`. But the resulting cut is not actually checked there.
Shall I add an output test of the cut in examples/simple ?

I just started at the sort-of top of the list, so if there are other tests which are a better starting point, please tell me.

**Comment by szhorvat** [comment]
@GroteGnoom Certainly, you can get started with `igraph_st_mincut`. Be sure to work on the `develop` branch ([instructions to compile](https://igraph.discourse.group/t/draft-document-for-installation-with-cmake/504))! There are two options: 

 - Add an _example_ in `examples/simple` which will be included in the documentation. This should be simple and educational.
 - Add a _test_ in `tests/unit` which doesn't have any requirements in simplicity.

If you need any help (how to compile, how to integrate tests, etc.), don't hesitate to ask. Please ask such questions on https://igraph.discourse.group/ and be sure to ping me as `@szhorvat` so I will be notified in email.

**Comment by GroteGnoom** [comment]
I made a wiki addition about adding tests here:
https://github.com/GroteGnoom/igraph/wiki/Adding-tests
But you can't really do wiki pull requests, and I also can't add new topics to the `discourse`. What's the right way to add things to the wiki?

**Comment by szhorvat** [comment]
> What's the right way to add things to the wiki?

Thanks so much for this! It's fine this way, and feel free to keep working on this wiki page. What I originally meant was that I would create the wiki page based on your feedback, but this is a perfect start. If it needs fixes, we can always edit and polish it later.

> and I also can't add new topics to the `discourse`.

That is unexpected. Can you open a new GitHub issue for that (for lack of a better discussion space) and describe what happens when you try to create a new topic?

**Comment by GroteGnoom** [comment]
igraph_write_graph_lgl can be crossed off the list because of #1601

**Comment by GroteGnoom** [comment]
#1622 fixes rewire_directed_edges
#1625 fixes lastcit_game

**Comment by GroteGnoom** [comment]
I am going to work on the communities, starting with igraph_compare_communities. Just for some variation :slightly_smiling_face:

**Comment by GroteGnoom** [comment]
igraph_vss_adj only has a declaration, but nothing else, so I think its declaration and its mention here can be removed.

**Comment by GroteGnoom** [comment]
I'll take a look at the iterators next.

**Comment by GroteGnoom** [comment]
igraph_sparsemat_nonzero_storage is tested in #1680
all 6 rng functions are tested in #1675
igraph_progress_handler_stderr and igraph_set_progress_handler are tested in #1674
 igraph_layout_drl is tested in #1670
igraph_layout_graphopt is tested in #1668
igraph_layout_random_3d is tested in #1667
 igraph_arpack_unpack_complex is tested in #1660
igraph_lapack_dgetrs is tested in #1657

**Comment by GroteGnoom** [comment]
igraph_progress_handler_stderr and igraph_set_progress_handler can be crossed off the list.

**Comment by szhorvat** [comment]
> Does that mean we should skip testing some of them?

Yes. It's good to prioritize. For example, I do not think that `igraph_adjlist_fprint` is very important.

> Does that sentence cover all the tests or the first list of tests with no header?

It covers all of them.

----

I originally came up with this list by going through the headers and looking at which functions my IDE is complaining about that they are "not used". "Not used" here means that there was no call to them in the entire codebase.

This is a good starting point to determine where the test coverage can be expanded, but this list should not be taken too rigidly. 

There are functions on this list which are indeed never called, but they are trivial enough that there's not much point in fully testing them (such as some of the printing functions). 

There are functions which are technically not called, but they are a thin wrapper for another function which is thoroughly exercised. 

There are functions which are complex and untested, but they are very rarely used and writing proper tests for them would take a great deal of work (including reading papers). One example is `igraph_graphlets`, which I am not 100% sure works correctly, but then it seems to be a very niche thing that was not generally adopted by the network science community ... One may debate if it is even a good analysis technique.

Then there are function which are not on this list because they happen to be called somewhere in the core codebase, but they do not have their own tests, or are not thoroughly tested, and it would be more useful to add tests for them than for maybe the majority of the functions which _are_ on the list.  Identifying these would take some work.  The coverage tools can be helpful to find functions where only some of the code paths are exercised.

**Comment by szhorvat** [comment]
Later I will add a `*` character next to a few of those functions in this list which I think would be nice to write tests for.  This is going to be somewhat subjective.

**Comment by GroteGnoom** [comment]
> This is going to be somewhat subjective though.

Of course! :slightly_smiling_face: But much better than not prioritizing. You can also be 'conservative', but then at least skip testing things like [zeroin](https://igraph.discourse.group/t/the-unit-tests-thread/597/27).

Knowing priorities is definitely nice, also for issues in general.

**Comment by szhorvat** [comment]
@GroteGnoom It looks like `igraph_transitivity_barrat` also suffers from assertion failures on the null graph. Are you planning to write tests for it within the upcoming days? If not, please let me know and I'll do it. If yes, please do this as well:

 - The "No weights given for Barrat's transitivity" warning should only be triggered if `no_of_edges > 0` (due to how some high-level interfaces work, there will not be any weight vector is the edge count is zero)
 - Check separately the case when there are zero vertices, and when there are some vertices, but the vertex selector passed to the function is empty.

**Comment by GroteGnoom** [comment]
Yes, I'll do that one soon 🙂

**Comment by GroteGnoom** [comment]
igraph_maximum_cardinality_search now seems pretty thoroughly tested because of the is_chordal tests, so I think it can be removed from the list. Or is there still a reason to test it extra carefully?
https://codecov.io/gh/igraph/igraph/src/master/src/misc/chordality.c
is_chordal itself apparently still has quite some gaps in it, I'll fix that.

**Comment by ntamas** [comment]
I think we can skip `igraph_maximum_cardinality_search`.

**Comment by szhorvat** [comment]
The SBM stuff may be good to test. It is fairly simple functionality, but it is also very important, and a candidate for future expansion. It's more important to test that than other random graph generators.

**Comment by Jeroenvh99** [comment]
I'd like to test igraph_bfgs, in which file is this function located?

**Comment by szhorvat** [comment]
First, a decision should be made whether we keep this function at all. It is not used anywhere within igraph, and not exposed in any high-level interfaces. It is not exactly within the scope of igraph, but I can imagine optimization being useful in the future.

@ntamas Opinions? Do you know why it was originally added? It seems to me that it's for a function that is now removed, see https://github.com/igraph/igraph/commit/c6fb3d4c0a4f240063b26c2ed21f4a7e09b0964b#diff-c9d46d321e819b992d0d7020669eedeea39ea459b71bc9e371a638f20f0a2058 and `igraph_revolver_ml_D()`

Note that this function was lifted from the R sources and then adapted to use igraph data structures.

If the decision is to keep this function, then the bigger part of the job is understanding and documenting the interface, concurrently within writing tests.

**Comment by szhorvat** [comment]
> in which file is this function located?

To answer this directly,

 - Use `grep` to search sources. There are also faster grep alternatives like ripgrep.
 - Use a good IDE that helps you navigate code. IMO this is a must for effective development. I use Qt Creator (free) and CLion (free for academic and open source use). Many use Visual Studio Code with appropriate extensions.

It's in `src/math/bfgs.c`

**Comment by Jeroenvh99** [comment]
I'll look into grep, thanks for your help!

**Comment by GroteGnoom** [comment]
I've removed the 'good first issue' label, because there are only special cases left, like undocumented or unnecessary functions.

**Comment by ntamas** [comment]
@szhorvat `igraph_bfgs` has been bothering me for a while; I thought that it was used somewhere, but I can't find any traces of that old function in the code so I'm going to remove it.

**Comment by szhorvat** [comment]
That's fine with me, but it _was_ public API. Are you sure you don't want to deprecate it instead?
